File: kotlin/wallet/src/main/kotlin/com/hopae/eudi/wallet/PresentationSession.kt
package com.hopae.eudi.wallet

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * A single presentation run driven as a state machine. One-shot; the flow runs
 * in the wallet scope and pauses at [PresentationState.RequestResolved] for the user's consent.
 */
class PresentationSession internal constructor(
    private val scope: CoroutineScope,
    private val flow: suspend PresentationSession.() -> Unit,
) {
    private val _state = MutableStateFlow<PresentationState>(PresentationState.ResolvingRequest)
    val state: StateFlow<PresentationState> = _state.asStateFlow()

    private var job: Job? = null
    private var decision: CompletableDeferred<PresentationSelection?>? = null

    internal fun launch() {
        job = scope.launch {
            try {
                flow()
            } catch (e: WalletError.Presentation) {
                _state.value = PresentationState.Failed(e)
            } catch (e: Throwable) {
                _state.value = PresentationState.Failed(WalletError.Presentation.Unexpected(e))
            }
        }
    }

    internal fun emit(state: PresentationState) {
        _state.value = state
    }

    /** Pauses at [PresentationState.RequestResolved] for consent; returns the selection or null (declined). */
    internal suspend fun awaitDecision(request: PresentationRequest): PresentationSelection? {
        val deferred = CompletableDeferred<PresentationSelection?>()
        decision = deferred
        emit(PresentationState.RequestResolved(request))
        return deferred.await()
    }

    /** Approve with the chosen credentials — resumes the flow to submit the response. */
    fun respond(selection: PresentationSelection) {
        decision?.complete(selection) ?: error("session is not awaiting a decision")
    }

    /** Decline the request — the flow terminates at [PresentationState.Declined]. */
    fun decline() {
        decision?.complete(null) ?: error("session is not awaiting a decision")
    }

    fun cancel() {
        job?.cancel()
    }
}

File: swift/Sources/Wallet/PresentationSession.swift
import Foundation

/**
 * A single presentation run driven as a state machine. One-shot; the flow runs
 * in a Task and pauses at `.requestResolved` for the user's consent until respond/decline.
 */
public final class PresentationSession: @unchecked Sendable {
    private let lock = NSLock()
    private var _currentState: PresentationState = .resolvingRequest
    public var currentState: PresentationState { locked { _currentState } }

    public let states: AsyncStream<PresentationState>
    private let continuation: AsyncStream<PresentationState>.Continuation
    private var task: Task<Void, Never>?
    private var decisionContinuation: CheckedContinuation<PresentationSelection?, Never>?
    private let flow: (PresentationSession) async throws -> Void

    init(_ flow: @escaping (PresentationSession) async throws -> Void) {
        self.flow = flow
        var c: AsyncStream<PresentationState>.Continuation!
        self.states = AsyncStream { c = $0 }
        self.continuation = c
    }

    func launch() {
        task = Task { [self] in
            do {
                try await flow(self)
            } catch let error as PresentationError {
                emit(.failed(error))
            } catch {
                emit(.failed(.unexpected(String(describing: error))))
            }
        }
    }

    func emit(_ state: PresentationState) {
        locked { _currentState = state }
        continuation.yield(state)
        if state.isTerminal { continuation.finish() }
    }

    /// Pauses at `.requestResolved` for consent; returns the selection or nil (declined).
    func awaitDecision(_ request: PresentationRequest) async -> PresentationSelection? {
        await withCheckedContinuation { cont in
            locked { decisionContinuation = cont } // set before emitting to avoid a resume-before-suspend race
            emit(.requestResolved(request))
        }
    }

    /// Approve with the chosen credentials — resumes the flow to submit the response.
    public func respond(_ selection: PresentationSelection) {
        resumeDecision(selection)
    }

    /// Decline the request — the flow terminates at `.declined`.
    public func decline() {
        resumeDecision(nil)
    }

    public func cancel() {
        task?.cancel()
        continuation.finish()
    }

    private func resumeDecision(_ value: PresentationSelection?) {
        let cont = locked { () -> CheckedContinuation<PresentationSelection?, Never>? in
            let c = decisionContinuation; decisionContinuation = nil; return c
        }
        cont?.resume(returning: value)
    }

    private func locked<T>(_ body: () -> T) -> T {
        lock.lock()
        defer { lock.unlock() }
        return body()
    }
}
